mempool/cat/cache.go

```diff
@@ -0,0 +1,292 @@
+	if c.size == 0 {
+		return false
+	}
+
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+
+	_, ok := c.cacheMap[txKey]
+	return ok
+}
+
+type EvictedTxInfo struct {
+	timeEvicted time.Time
+	priority    int64
+	gasWanted   int64
+	sender      string
+	size        int64
+}
+
+type EvictedTxCache struct {
+	mtx   tmsync.Mutex
+	size  int
+	cache map[types.TxKey]*EvictedTxInfo
+}
+
+func NewEvictedTxCache(size int) *EvictedTxCache {
+	return &EvictedTxCache{
+		size:  size,
+		cache: make(map[types.TxKey]*EvictedTxInfo),
+	}
+}
+
+func (c *EvictedTxCache) Has(txKey types.TxKey) bool {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	_, exists := c.cache[txKey]
+	return exists
+}
+
+func (c *EvictedTxCache) Get(txKey types.TxKey) *EvictedTxInfo {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	return c.cache[txKey]
+}
+
+func (c *EvictedTxCache) Push(wtx *wrappedTx) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	c.cache[wtx.key] = &EvictedTxInfo{
+		timeEvicted: time.Now().UTC(),
+		priority:    wtx.priority,
+		gasWanted:   wtx.gasWanted,
+		sender:      wtx.sender,
+		size:        wtx.size(),
+	}
+	// if cache too large, remove the oldest entry
+	if len(c.cache) > c.size {
+		oldestTxKey := wtx.key
+		oldestTxTime := time.Now().UTC()
+		for key, info := range c.cache {
+			if info.timeEvicted.Before(oldestTxTime) {
+				oldestTxTime = info.timeEvicted
+				oldestTxKey = key
+			}
+		}
+		delete(c.cache, oldestTxKey)
+	}
+}
+
+func (c *EvictedTxCache) Pop(txKey types.TxKey) *EvictedTxInfo {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	info, exists := c.cache[txKey]
+	if !exists {
+		return nil
+	} else {
+		delete(c.cache, txKey)
+		return info
+	}
+}
+
+func (c *EvictedTxCache) Prune(limit time.Time) {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	for key, info := range c.cache {
+		if info.timeEvicted.Before(limit) {
+			delete(c.cache, key)
+		}
+	}
+}
+
+func (c *EvictedTxCache) Reset() {
+	c.mtx.Lock()
+	defer c.mtx.Unlock()
+	c.cache = make(map[types.TxKey]*EvictedTxInfo)
+}
+
+// SeenTxSet records transactions that have been
+// seen by other peers but not yet by us
+type SeenTxSet struct {
+	mtx tmsync.Mutex
+	set map[types.TxKey]timestampedPeerSet
+}
+
+type timestampedPeerSet struct {
+	peers map[uint16]bool
+	time  time.Time
+}
+
+func NewSeenTxSet() *SeenTxSet {
+	return &SeenTxSet{
+		set: make(map[types.TxKey]timestampedPeerSet),
+	}
+}
+
+func (s *SeenTxSet) Add(txKey types.TxKey, peer uint16) {
+	if peer == 0 {
+		return
+	}
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	seenSet, exists := s.set[txKey]
+	if !exists {
+		s.set[txKey] = timestampedPeerSet{
+			peers: map[uint16]bool{peer: true},
+			time:  time.Now().UTC(),
+		}
+	} else {
+		seenSet.peers[peer] = true
+	}
+}
+
+func (s *SeenTxSet) Pop(txKey types.TxKey) uint16 {
+	s.mtx.Lock()
+	defer s.mtx.Unlock()
+	seenSet, exists := s.set[txKey]
+	if !exists {
+		return 0
+	} else {
+		for peer := range seenSet.peers {
+			delete(seenSet.peers, peer)
+			return peer
+		}
+		return 0
+	}
```

Comment:
nit: I think it is cleaner to write
```suggestion
	if exists {
		for peer := range seenSet.peers {
			delete(seenSet.peers, peer)
			return peer
		}
	}
	return 0
```